### PR TITLE
Allow derived types to override intrinsic parent types

### LINKED ISSUE #1137: Allow overriding of intrinsic type property in derived types
From design issue [#762](https://github.com/microsoft/cadl/issues/762)

Allow the following

```
model Parent {
  name: string;
}

model Child extends Parent {
  name: "abc"
}
```

Rule for allowing override:
- base type MUST be a intrinsic type
- derived type MUST be a subtype of the base type

### PR DESCRIPTION
**Pull request by tjprescott**
Fixes #1137.

Please take a hard look at the tests and let me know if there are any test cases I should add. This issue felt easy... yeah, a little too easy.

### CODE REVIEW

File: packages/compiler/core/checker.ts [hunk 1]
```diff
@@ -2136,12 +2136,13 @@ export function createChecker(program: Program): Checker {
       type.namespace?.models.set(type.name, type);
     }
 
-    const inheritedPropNames = new Set(
-      Array.from(walkPropertiesInherited(type)).map((v) => v.name)
-    );
+    const inheritedProperties = new Map<string, ModelProperty>();
+    Array.from(walkPropertiesInherited(type)).forEach((item) => {
```
Comment (bterlson) [inline packages/compiler/core/checker.ts:2140]: Does this work right when the base model shadows a property from superclass? I suppose due to the transitive property of the subtype relationship it should, but it seems like properties higher up the base class list will "win" in this list which feels risky.

File: packages/compiler/core/checker.ts [hunk 2]
```diff
@@ -2265,15 +2266,19 @@ export function createChecker(program: Program): Checker {
...
-        createDiagnostic({
-          code: "override-property",
-          format: { propName: newProp.name },
-          target: diagnosticTarget ?? newProp,
-        })
-      );
-
+    if (inheritedProperties?.has(newProp.name)) {
+      const inheritedProp = inheritedProperties.get(newProp.name)!;
+      const [isAssignable, _] = isTypeAssignableTo(newProp.type, inheritedProp.type, newProp);
```
Comment (bterlson) [inline packages/compiler/core/checker.ts:2271]: not part of this PR but I don't like "assignable" since our specs I think only care about subtyping, we don't have a notion of assignability yet.

File: packages/compiler/core/checker.ts [hunk 3]
```diff
@@ -2265,15 +2266,19 @@ export function createChecker(program: Program): Checker {
...
+      const parentIntrinsic = isIntrinsic(program, inheritedProp.type);
```
Comment (bterlson) [inline packages/compiler/core/checker.ts:2272]: Doesn't this say that we allow any type at all to override parent intrinsic types?

File: packages/compiler/core/checker.ts [hunk 4]
```diff
@@ -2265,15 +2266,19 @@ export function createChecker(program: Program): Checker {
...
+      if (!parentIntrinsic || !isAssignable) {
+        program.reportDiagnostic(
+          createDiagnostic({
+            code: "override-property",
+            format: { propName: newProp.name },
```
Comment (bterlson) [inline packages/compiler/core/checker.ts:2277]: I think this diagnostic needs to be updated given that you can override a property with this PR (it just needs to be a subtype of the parent type).

Comment (bterlson) [inline packages/compiler/core/checker.ts:2277]: Need to update the diagnostic message.

File: packages/compiler/core/checker.ts (hunk 3 above)
Comment (bterlson) [reply packages/compiler/core/checker.ts:2272]: Sorry I see, the intention is that we disallow subtyping of non-intrinsic types. Is that what we agreed to in a design discussion?

File: packages/compiler/core/checker.ts (hunk 1 above)
Comment (tjprescott) [reply packages/compiler/core/checker.ts:2140]: Could you provide an example? Are you talking about something like A extends B and C extends B?

File: packages/compiler/core/checker.ts (hunk 3 above)
Comment (tjprescott) [reply packages/compiler/core/checker.ts:2272]: Yes, @timotheeguerin calls it out here: https://github.com/microsoft/cadl/issues/1137

If the derived property isn't assignable to the parent property type or it is but the parent property type isn't intrinsic, it fails.

File: packages/compiler/core/checker.ts (hunk 4 above)
Comment (tjprescott) [reply packages/compiler/core/checker.ts]: I split the diagnostic into two new messages.

File: packages/compiler/core/checker.ts (hunk 1 above)
Comment (bterlson) [reply packages/compiler/core/checker.ts:2140]: Yeah exactly, something like:

```
model A {
  prop: Type
}

model B extends A {
  prop: Subtype
}

model C extends B {
  prop: SubSubtype
}
```

I think this algorithm will compare against `Type` not `Subtype` when asking if C's prop is a valid subtype.

File: packages/compiler/core/checker.ts (hunk 3 above)
Comment (bterlson) [reply packages/compiler/core/checker.ts:2272]: ok seems good for now.

File: packages/compiler/core/checker.ts (hunk 4 above)
Comment (bterlson) [reply packages/compiler/core/checker.ts]: nice looks good. The error messages are more verbose than I would like but this is true of many so I'm ok cleaning this up in a pass later.

File: packages/compiler/core/checker.ts (hunk 1 above)
Comment (tjprescott) [reply packages/compiler/core/checker.ts:2140]: I created this relation:
```
model A {
  prop: int64;
}

model B extends A {
  prop: int32;
}

model C extends B {
  prop: int16;
}
```

This works, as I would expect, but if I change B.prop to say `int8`, the compiler is still happy, even though I would expect it to fail (since int16 won't fit in int8). However, it's clear that the compiler is comparing C.prop (int16) to A.prop (int64). Is this the scenario you are concerned with?

Comment (tjprescott) [reply packages/compiler/core/checker.ts:2140]: > Yeah that's it in general (though I'm unsure if we have any subtype relationships between the sized integer types).

I think we must because if I try to make A.prop int8 it does fail.

File: packages/compiler/core/checker.ts (hunk 2 above)
Comment (timotheeguerin) [reply packages/compiler/core/checker.ts]: From what I remember this was on how the typescript checker did the type relation it had a `isTypeRelatedTo` function and you'd pass the relation type(inheritence, assignment, etc.) not sure what other there was but assignment felt like the correct description

File: packages/compiler/core/checker.ts [hunk 5]
```diff
@@ -2265,16 +2264,32 @@ export function createChecker(program: Program): Checker {
...
-        createDiagnostic({
-          code: "override-property",
-          format: { propName: newProp.name },
-          target: diagnosticTarget ?? newProp,
-        })
-      );
-
-      return;
+    if (inheritedProperties?.has(newProp.name)) {
+      for (const inheritedProp of inheritedProperties.get(newProp.name)!) {
```
Comment (timotheeguerin) [inline packages/compiler/core/checker.ts:2268]: nit:
instead of `inheritedProperties?.has(newProp.name)` then `inheritedProperties.get(newProp.name)!` you can just assign the result of `.get` to a variable and check the if there. That make it so you don't have to force saying `.get` is not `undefined`

**Comment by tjprescott** [comment]
@bterlson I think all of your comments have been addressed.

**Comment by tjprescott** [comment]
Thanks for the review @timotheeguerin. I've incorporated your comments in the latest commit.

File: packages/compiler/core/checker.ts [hunk 6]
```diff
@@ -2136,12 +2136,8 @@ export function createChecker(program: Program): Checker {
       type.namespace?.models.set(type.name, type);
     }
 
-    const inheritedPropNames = new Set(
-      Array.from(walkPropertiesInherited(type)).map((v) => v.name)
```
Comment (nguerrera) [inline packages/compiler/core/checker.ts:2140]: (Not related to wanting this particular line of code changed.)

I think walkPropertiesInherited may still need to be changed. I don't expect that callers are prepared to see something and its override(s) returned. I think it should probably not return properties that are overridden, at least not by default. I could see an opt-in arg to include them so that you signal that you understand that.

Here's one example: https://github.com/microsoft/cadl/blob/c5a2a75fed07dd0aa3078b244749331a7b7adccc/packages/rest/src/http/metadata.ts#L156

Looks like if you have, for example if you have a base property marked `@header` and an override that isn't marked as `@header` you will still get the header defined by the base, which seems wrong.


There's also this where I anticipated an issue in getEffectiveType once overrides were possible:

https://github.com/microsoft/cadl/blob/c5a2a75fed07dd0aa3078b244749331a7b7adccc/packages/compiler/core/checker.ts#L4577-L4580

It also looks like member references might do the wrong thing in lateBindMembers by binding inherited over top of overridden. However, I couldn't repro this in the playground, so there's probably something wrong with my reasoning and I'm investigating that.

Comment (nguerrera) [reply packages/compiler/core/checker.ts:2140]: lateBindMembers is working nearly by accident. It goes to replace the override with the base property but that just adds to a duplicate list that is never reported as a diagnostic since it happens in late binding. This adds to my intuition that walkPropertiesInherited should be changed to exlude base properties that are overridden. And it would be great to try to add test cases that have overrides that cover paths through walkPropertiesInherited.

Comment (tjprescott) [reply packages/compiler/core/checker.ts:2140]: Yes, `walkInheritedProperties` didn't work at all like I expected it to, which is why I stopped using it. In the case of A extends B extends C, it returns A.prop when called on C even if B also overrides prop. The current logic I implemented simply traverses the inheritance chain and finds the first override for a given property, returning that. 

I don't understand any of the late binding logic, now what the purpose of walkInhertiedProperties is supposed to be, so I left all of that alone.

Comment (nguerrera) [reply packages/compiler/core/checker.ts:2140]: walkInheritedProperties simply enumerates all of the properties in the inheritance chain (with most derived enumerated before least derived). It is used whenever something wants to loop over all properties of a model including inherited properties. Until this change, overrides were impossible so the implementation was trivial. Now we have to decide what it should do when there are overrides. It's not enough that you could implement this change withou walkInheritedProperties, we have to consider all the other existing callers to walkInheritedProperties. And then we need to either change walkInheritedProperties or change the callers if overrides cause them issues.

Looking at just a few calls to walkInheritedProperties, I can find bugs:

1. Override a `@header` with a non-header and the base header remains (could also be `@path`, `@query`, etc:

https://cadlplayground.z22.web.core.windows.net/prs/1248/?c=aW1wb3J0ICJAY2FkbC1sYW5nL3Jlc3QiOwoKQHNlcnZpY2UoewogIHRpdGxlOiAiV2lkZ2V0IFPGGyIsCn0pCm5hbWVzcGFjZSBEZW1vxxs7Cgp1c2luZyBDYWRsLkh0dHA7Cgptb2RlbCBCYXNlIMRVcHJvcDogc3RyaW5nOwp9yCDHZGV4dGVuZHPQLyJ05gCgfQoKb3AgxA0oKTrINCYge307

2. Cannot intersect or a model with a property override with anything

https://cadlplayground.z22.web.core.windows.net/prs/1248/?c=aW1wb3J0ICJAY2FkbC1sYW5nL3Jlc3QiOwoKQHNlcnZpY2UoewogIHRpdGxlOiAiV2lkZ2V0IFPGGyIsCn0pCm5hbWVzcGFjZSBEZW1vxxs7Cgp1c2luZyBDYWRsLkh0dHA7Cgptb2RlbCBCYXNlIMRVQGhlYWRlciBoMTogc3RyaW5nOwp9yCbHamV4dGVuZHPKNcQtInTmAKR9CgpvcCDEDSgpOscyOw%3D%3D

3. Cannot spread a model a model with a property override

https://cadlplayground.z22.web.core.windows.net/prs/1248/?c=aW1wb3J0ICJAY2FkbC1sYW5nL3Jlc3QiOwoKQHNlcnZpY2UoewogIHRpdGxlOiAiV2lkZ2V0IFPGGyIsCn0pCm5hbWVzcGFjZSBEZW1vxxs7Cgp1c2luZyBDYWRsLkh0dHA7Cgptb2RlbCBCYXNlIMRVQGhlYWRlciBoMTogc3RyaW5nOwp9yCbHamV4dGVuZHPKNcQtInTmAKR9CgpvcCDEDSgpOiB7IC4uLsc3fTs%3D

I suspect there is also a bug in getEffectiveType in the location I mentioned where I say this will need to change once overrides arrive, but (2) and (3) got in the way of reproing it.

lateBindMembers is what makes it possible to reference a member, for example you can reference property P of model M using `P.M`. There isn't an observable bug that I can find there, but if a property is overridden, it is getting lucky that the symbol table is recording a duplicate instead of overwriting.

So far in all these cases, I believe what the caller wants is for the overridden members to be excluded from being enumerated.

Comment (tjprescott) [reply packages/compiler/core/checker.ts:2140]: I'll convert these to test cases.

Comment (tjprescott) [reply packages/compiler/core/checker.ts:2140]: So, regarding item 1, is there a general rule? I don't think we are saying that an overridden property needs to have the same decorators as the base property, or are we? Should this just be a separate linter rule in the Http library?

The other two seem straightforward to me. They should allow the intersection or spreading with the derived property, not the base.

Comment (nguerrera) [reply packages/compiler/core/checker.ts:2140]: I think it would be fine if the header disappeared in (1) which is what would happen if walkInheritedProperties didn't return overridden properties. I believe this same change would fix all 3.

Comment (nguerrera) [reply packages/compiler/core/checker.ts:2140]: I think the general rule is an override completely replaces the base property. If it drops a decorator, then you get the consequences. But now what's happening is both the base and the override are contributing to the swagger and not in any coherent way.

Comment (tjprescott) [reply packages/compiler/core/checker.ts:2140]: Thanks @nguerrera. I added those test cases and updated the logic of walkPropertiesInherited to not return duplicates (it will return the closest relative). Please try out those other scenarios you mentioned and let me know if anything else needs changing.

Comment (tjprescott) [reply packages/compiler/core/checker.ts:2140]: @nguerrera please take a second look.

Comment (nguerrera) [reply packages/compiler/core/checker.ts:2140]: I will look at this first thing in the morning. Sorry that I forgot.

Comment (nguerrera) [reply packages/compiler/core/checker.ts:2140]: I have been looking at this all morning. The effective type issue is not easy to describe or fix. I will share a commit with test cases and the fix when I have it.

Comment (nguerrera) [reply packages/compiler/core/checker.ts:2140]: I sent https://github.com/tjprescott/cadl/pull/1 with the changes I suggest.

File: packages/compiler/core/checker.ts [hunk 7]
```diff
@@ -4658,26 +4671,54 @@ export function filterModelProperties(
...
+    property.model,
+    "Parent model must be set before overridden property can be found."
+  );
+
+  for (let current = property.model.baseModel; current; current = current.baseModel) {
+    const overriden = current.properties.get(property.name);
+    if (overriden) {
+      return overriden;
+    }
+  }
```
Comment (nguerrera) [inline packages/compiler/core/checker.ts:4689]:
```suggestion
  for (let current = property.model.baseModel; current; current = current.baseModel) {
    const overridden = current.properties.get(property.name);
    if (overridden) {
      return overridden;
    }
  }
```

Looks like I left you a spell check failure, oops.

**Review by nguerrera — APPROVED** [review]
